models/review.go

```diff
@@ -345,9 +345,9 @@ func GetReviewersByPullID(pullID int64) (issueReviewers []*PullReviewersWithType
 	irs := []*PullReviewersWithType{}
 	if x.Dialect().DBType() == core.MSSQL {
 		err = x.SQL(`SELECT [user].*, review.type, review.official, review.review_updated_unix FROM
-(SELECT review.id, review.type, review.reviewer_id, max(review.updated_unix) as review_updated_unix
+(SELECT review.id, review.type, review.reviewer_id, review.official, max(review.updated_unix) as review_updated_unix
```

Comment:
This line I don't understand. If you have `review.id` in the row, all lines are already unique (unless there were multiple users with the same id which should never happen). A `MAX(updated_unix)` will be the same as `updated_unix` without the Group By. 🤔

What is it that you are trying to accomplish? Bring up the latest review? The query from the migration should achieve that.

Reply:
The function shall return the latest review of each reviewer (not necessary official reviewers). I see that I actually don't use the change in this function, probably rewrote something. I think later on the information on which reviews are "official" should be shown on the pull request page.

The `max(review.updated_unix) as review_updated_unix` is used to sort the list and ensure the group by gets the latest review. However it might have been enough to get the `max(id)` instead, I don't see how reviews from same reviewer could change order.

Reply:
> 
> 
> The function shall return the latest review of each reviewer (not necessary official reviewers). I see that I actually don't use the change in this function, probably rewrote something. I think later on the information on which reviews are "official" should be shown on the pull request page.
> 
> The `max(review.updated_unix) as review_updated_unix` is used to sort the list and ensure the group by gets the latest review. However it might have been enough to get the `max(id)` instead, I don't see how reviews from same reviewer could change order.

This kind of query will not work. If you have two different `review.type` (e.g. comment + approval) from the same user, you'll get two rows because Group By will see them as different values. To get the latest record you need to do a select in select like the one used in the migration:

```golang
if err := sess.SQL("SELECT * FROM review WHERE id IN (SELECT max(id) as id FROM review WHERE issue_id = ? AND type in (?, ?) GROUP BY issue_id, reviewer_id)",
			issueID, models.ReviewTypeApprove, models.ReviewTypeReject).
			Find(&reviews); err != nil {
			return err
		}
```

Of course the example above requires some adaptation to get the JOIN and the columns you're expecting.

Reply:
It is actually solved further down in the function (I don't say it is a good solution 😄)

```
	// We need to group our results by the fields we want, otherwise some databases will fail (only_full_group_by).
	// But becaus we're doing this, we need to manually filter out multiple reviews of different types by the
	// same person because we only want to show the newest review grouped by user. Thats why we're using a map here.
```

Reply:
Your query might work... it doesn't work to just remove some of the group by fields for pgsql and mysql (due to some _only_full_group_by_ setting).

Reply:
So I didn't want to touch this function more than needed (and actually I can revert the changes I done). But the only thing I added was the `Official` field to the results, and hence added in SQL where necessary. The multiple rows is handled by the code later already. 

I think maybe this function tries to do too much in SQL which makes it a bit complex.

Reply:
To exemplify this, let me give you one example. Suppose you have this in the `review` table:

| review ID  | user_id     | review_type         | review_date         |
|------------|-------------|---------------------|--------------|
| 1          | 1           | comment             | Jan 1        |
| 2          | 2           | request-changes     | Jan 2        |
| 3          | 1           | comment             | Jan 3        |
| 4          | 3           | comment             | Jan 4        |
| 5          | 2           | approval            | Jan 5        |

If you do:

```SQL
select max(review_id), user_id, review_type, max(review_date)
from review
group by user_id, review_type;
```

This is what you will get:

| review ID  | user_id     | review_type         | review_date         |
|------------|-------------|---------------------|--------------|
| 2          | 2           | request-changes     | Jan 2        |
| 3          | 1           | comment             | Jan 3        |
| 4          | 3           | comment             | Jan 4        |
| 5          | 2           | approval            | Jan 5        |

As you see, the two reviews from user `#1` were consolidated in one row, as the fields `user_id` and `review_type` are the same for all records; user `#3` is not a problem, but the two reviews from user `#2` show up as different rows.

That's why the only way to get this SQL right is with a sub-query:

```SQL
select max(id), user_id from review group by user_id;
```

Will give you:

| review ID  | user_id     |
|------------|-------------|
| 3          | 1           |
| 4          | 3           |
| 5          | 2           |

So, these are the records that you're looking for:

```SQL
select review_id, user_id, review_type, review_date
from `review`
where review_id in (
        select max(id), user_id from `review` as `review_inner` group by user_id
    )
```

returns:

| review ID  | user_id     | review_type         | date         |
|------------|-------------|---------------------|--------------|
| 3          | 1           | comment             | Jan 3        |
| 4          | 3           | comment             | Jan 4        |
| 5          | 2           | approval            | Jan 5        |

which is what you want. With this method you can perform joins too:

```SQL
select review_id, user_id, review_type, review_date, `user`.*
from `review`
INNER JOIN `user` ON `user`.`id` = `review`.`user_id`
where review_id in (
        select max(id), user_id from `review` as `review_inner` group by user_id
    )
```

There's probably a nice way to write all of this with a couple of xorm directives (e.g. I think you can write the subquery in builder and use it in the `In()` operator, but I'm not sure), so I think it will be just quicker to do it with `sess.SQL()` like you did in migrations.

Reply:
Very long informative response 😄. 
I just wanted to point out that the existing code actually works, although not very efficient or elegant. I will see if I can rewrite the SQL query in a better way.

Reply:
Oh, sorry! I didn't realize that the code wasn't yours. It's my tendency of checking the right column more thoroughly than the left one. 😋

If you just don't feel like doing that change just let me know, because that's the only thing I was waiting for to give my 👍.

Reply:
I rewrote the code and think it is easier to understand now although there will be a few more sql queries, but normally there will not be that many reviews for a PR.